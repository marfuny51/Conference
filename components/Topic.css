.Table {
    text-align: center;
}

.TableTopic {
    border-collapse: collapse;
    width: 70vw;
    height: 10vh;
    margin: 6vh auto 2vh auto;
    table-layout: fixed;
}

.TableTopicAdd {
    border-collapse: collapse;
    width: 70vw;
    height: 10vh;
    margin: 6vh auto 2vh auto;
    opacity:0.3;
    table-layout: fixed;
}

.TableHead {
    color: #00D970;
}

.TableHead th {
    font-family: 'Muli', sans-serif;
    font-size: 1.2vw;
    background-color: #127a16;
    color: white;
    letter-spacing: 0.1vw;
    padding: 1vw;
    border: 0vw;
}

.TableBody tr {
    height: 5vw;
    background-color: #d3f1d4;
    border-bottom: 0.15vw solid #127a16;
}

.TableBody tr:hover {
    background-color: #f5f5f5;
}

.TableBody tr td {
    width: 13vw;
    font-family: 'Muli', sans-serif;
    font-size: 1.1vw;
    border: 0vw;
    padding: 0.5vw;
}

.Read {
    font-family: 'Muli', sans-serif;
    font-size: 1vw;
    height: 2vw;
    background-color: #a2eca3;
    color: black;
    border-radius: 0.5vw;
    border:0.2vw solid #77b45b;
}

.Read:hover {
    font-weight: bold;
    border:0.2vw solid #456835;
    cursor: pointer;
    transform: scale(1.03);
}

.ReadNone {
    display: none;
}

.Read2 {
    font-family: 'Muli', sans-serif;
    font-size: 1vw;
    height: 2vw;
    background-color: #a2eca3;
    color: black;
    border-radius: 0.5vw;
    border:0.2vw solid #77b45b;
    margin-right: 1.2vw;
}

.Read2:hover {
    border:0.2vw solid #456835;
    cursor: pointer;
    transform: scale(1.03);
}

.Add {
    font-family: 'Muli', sans-serif;
    font-size: 1.1vw;
    font-weight: bold;
    height: 3vw;
    background-color: #a2eca3;
    color: black;
    border-radius: 0.5vw;
    border:0.2vw solid #77b45b;
}

.Add:hover {
    transform: scale(1.03);
    border:0.2vw solid #456835;
    cursor: pointer;
}

.NavLink {
    color: #172c11;
    font-weight: bold;
}

.LoadData {
    font-size: 1.2vw;
    text-align: center;
    color: white;
}

.AddTopic {
    width: 48vw;
    height: 24vw;
    position: absolute;
    top: 35%;
    left: 26%;
    background-color: #d3f1d4;
    border: 0.2vw solid #77b45b;
    border-radius: 0.5vw;
}

.Title {
    font-family: 'Muli', sans-serif;
    font-weight: bold;
    letter-spacing:0.1vw;
    padding-top: 1vw;
}

.Input {
    font-family: 'Muli', sans-serif;
    font-size: 1.1vw;
    text-align: left;
    padding: 3.8vh 1vw 3vh 4vw;
}

.Input input{
    font-size: 1vw;
    width: 9vw;
    height: 1.2vw;
}

.Input textarea{
    font-size: 1vw;
    width: 16vw;
    height: 10vw;
}

.Valide {
    color: red;
}

.Textarea {
    width: 18vw;
    height: 18vh;
}

.Hide {
    width: 6.2vw;
    font-family: 'Muli', sans-serif;
    font-size: 1vw;
    background-color: #2f4f30;
    color: white;
    border-radius: 0.5vw;
    border:0.2vw solid #77b45b;
    margin-bottom: 1vh;
}

.Hide:hover {
    border:0.2vw solid #d3f1d4;
    cursor: pointer;
    transform: scale(1.03);
}

.ReadDiv {
    width: 27vw;
    height: 0vw;
    margin: auto;
    opacity: 0;
    display: none;
}

.ReadDivAnim1 {
    display: block;
    transition-duration: 0.7s;
    -webkit-transition-timing-function: linear; 
    transition-timing-function: linear;
    overflow: hidden;
    width: 27vw;
    height: 10vw; 
    opacity: 1;  
}

.ReadDivAnim2 {
    transition-duration: 0.7s;
    -webkit-transition-timing-function: linear; 
    transition-timing-function: linear;
    overflow: hidden;
    width: 27vw;
    height: 0vw; 
    opacity: 0;  
}

.TableBody .LinkTd {
    font-weight: bold;
}

.col1 {
    width: 15vw;
}

.col2 {
    width: 10vw;
}

.col3 {
    width: 27vw;
}

.col4 {
    width: 6vw;
}