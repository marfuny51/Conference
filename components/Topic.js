import React from 'react';
import PropTypes from 'prop-types';
import { NavLink } from 'react-router-dom';
import {voteEvents} from './events';

import './Topic.css';

class Topic extends React.PureComponent {

  static propTypes = {
    topic:PropTypes.shape({
      id: PropTypes.number.isRequired,
      title: PropTypes.string.isRequired,
      mainWords: PropTypes.string.isRequired,
      author: PropTypes.string.isRequired,
      info: PropTypes.string.isRequired,
    }),
  };

  deleteTopic = () => {
    voteEvents.emit('EDeleteTopic', this.props.topic.id);
  }

  editTopic = () => {
    voteEvents.emit('EEditTopic', this.props.topic.id);
  }

  moreRead = () => {
    voteEvents.emit('EReadMore', this.props.topic.id);
  }

  render() {
    let info = <div className='ReadDiv'>{this.props.topic.info}<br/><input type="button" value="Hide" className='Read' onClick={this.hide}/></div>; 
    let readMore = <input type="button" value="Read more..." className='Read' onClick={this.moreRead}/>;
    return (
      <tr key={this.props.topic.id}>
        <td><NavLink to={"/topic/"+this.props.topic.title} className='NavLink'>{this.props.topic.title}</NavLink></td>
        <td>{this.props.topic.mainWords}</td>
        <td>{this.props.topic.author}</td>
        <td>{(this.props.mode===55)?info:readMore}</td>
        <td><input type="button" value="Edit" className='Read' disabled = {(this.props.mode===2)?true:false} onClick={this.editTopic}/></td>
        <td><input type="button" value="Delete" className='Read' disabled = {(this.props.mode===2)?true:false} onClick={this.deleteTopic}/></td>
      </tr>
    );

  }

}

export default Topic;
