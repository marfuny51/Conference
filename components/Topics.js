import React from 'react';
import PropTypes from 'prop-types';
import {voteEvents} from './events';
import isoFetch from 'isomorphic-fetch';
import Topic from './Topic';
import AddTopic from './addTopic';
import EditSpeaker from './editSpeaker';

import './Topics.css';

class Topics extends React.PureComponent {

  /*constructor(props) {
    super(props);
    this.mounted = false;
  }*/

  state = {
    dataReady: false,
    topics: [],
    deleteCode: null,
    editCode: null,
    mode: null, //3- edit, 4 -add, 5- view
  }

  topicsArray;

  componentDidMount = () => {
    this.loadData();
    //this.mounted = true;
    voteEvents.addListener('ESave',this.topicSave);
    voteEvents.addListener('ECancel',this.topicCancel);
    voteEvents.addListener('EDelete',this.topicDelete);
    voteEvents.addListener('EEdit',this.editTopic);
    voteEvents.addListener('EditSave',this.topicSave);     
  };

  componentWillUnmount = () => {
    //this.mounted = false;
    voteEvents.removeListener('ESave',this.topicSave);
    voteEvents.removeListener('ECancel',this.topicCancel);
    voteEvents.removeListener('EDelete',this.topicDelete);
    voteEvents.removeListener('EEdit',this.editTopic);
    voteEvents.removeListener('EditSave',this.topicSave);
  };

  loadData = () => {
    let ajaxHandlerScript="https://fe.it-academy.by/AjaxStringStorage2.php";
    let sp = new URLSearchParams();
    sp.append('f', 'READ');
    sp.append('n', 'PROBA_PERA7');

    isoFetch(ajaxHandlerScript, {
        method: 'post',
        headers: {
            "Accept": "application/json",
        },
        body: sp,
    })
        .then( response => { 
            if (!response.ok)
                throw new Error("fetch error " + response.status); 
            else
                return response.json(); 
        })
        .then( data => {
            if (data.result === "") {
              this.speakersArray = {};
              this.speakersArray.speakers = [];
              this.speakersArray.topics = [];
              this.speakersArray.members = [];
            }
            else this.speakersArray = JSON.parse(data.result);
            console.log(this.speakersArray.topics);
            
              this.setState({ dataReady:true, topics:this.speakersArray.topics});
            
        })
        .catch( error => {
            this.fetchError(error.message);
        });

  }; 

  fetchError = (errorMessage) => {
    console.error(showStr);
  };

  topicSave = (id, title, mainWords, author, info) => {
    var ajaxHandlerScript="https://fe.it-academy.by/AjaxStringStorage2.php";
    var updatePassword=Math.random();
    let sp1 = new URLSearchParams();
    sp1.append('f', 'LOCKGET');
    sp1.append('n', 'PROBA_PERA7');
    sp1.append('p', updatePassword);
    
    isoFetch(ajaxHandlerScript, {
      method: 'post',
      headers: {
          "Accept": "application/json",
      },
      body: sp1,
    })
      .then( response => { 
          if (!response.ok)
              throw new Error("fetch error " + response.status); 
          else
              return response.json(); 
      })
      .then( data => {
          if (data.result === "") {
            this.speakersArray = {};
            this.speakersArray.topics = [];
          }
          else this.speakersArray.topics = JSON.parse(data.result.topics);
      })
      .catch( error => {
          console.log(error.message);
      });
    
      if (this.state.mode === 3) {
        console.log(title, mainWords, author, info);
        this.speakersArray.topics.forEach( topic => {
          if (topic.id === id) {
            topic.title = title;
            topic.mainWords = mainWords;
            topic.author = author;
            topic.info = info;
          }
        })
      }

      else if (this.state.mode === 4) {
        this.speakersArray.topics.push({id: id, title: title, mainWords: mainWords, author: author, info: info});
        console.log(this.speakersArray);
      }

    let sp2 = new URLSearchParams();
    sp2.append('f', 'UPDATE');
    sp2.append('n', 'PROBA_PERA7');
    sp2.append('p', updatePassword);
    sp2.append('v', JSON.stringify(this.speakersArray));

    isoFetch(ajaxHandlerScript, {
      method: 'post',
      headers: {
          "Accept": "application/json",
      },
      body: sp2,
    })
      .then( response => { 
          if (!response.ok)
              throw new Error("fetch error " + response.status); 
          else
              return response.json(); 
      })
      .then( (data) => {
        console.log(data);
        this.setState({mode:5});
        
          this.loadData();
        
      })
      .catch( error => {
          this.fetchError(error.message);
      });
    }

    delete = (id) => {
      var ajaxHandlerScript="https://fe.it-academy.by/AjaxStringStorage2.php";
      var updatePassword=Math.random();
      let sp3 = new URLSearchParams();
      sp3.append('f', 'LOCKGET');
      sp3.append('n', 'PROBA_PERA7');
      sp3.append('p', updatePassword);
      
      isoFetch(ajaxHandlerScript, {
        method: 'post',
        headers: {
            "Accept": "application/json",
        },
        body: sp3,
      })
        .then( response => { 
            if (!response.ok)
                throw new Error("fetch error " + response.status); 
            else
                return response.json(); 
        })
        .then( data => {
            if (data.result === "") {
              this.speakersArray = {};
              this.speakersArray.topics = [];
            }
            else this.speakersArray = JSON.parse(data.result);
        })
        .catch( error => {
            console.log(error.message);
        });
      
      this.speakersArray.topics = this.speakersArray.topics.filter(topic => topic.id !== id);
      console.log(this.speakersArray);
  
      let sp2 = new URLSearchParams();
      sp2.append('f', 'UPDATE');
      sp2.append('n', 'PROBA_PERA7');
      sp2.append('p', updatePassword);
      sp2.append('v', JSON.stringify(this.speakersArray));
  
      isoFetch(ajaxHandlerScript, {
        method: 'post',
        headers: {
            "Accept": "application/json",
        },
        body: sp2,
      })
        .then( response => { 
            if (!response.ok)
                throw new Error("fetch error " + response.status); 
            else
                return response.json(); 
        })
        .then( (data) => {
          
            this.loadData();
          
        })
        .catch( error => {
            this.fetchError(error.message);
        });
      }

  topicCancel = () => {
    let topics = [...this.state.topics];
    this.setState({mode:5, topics: topics});
  }

  addTopic = (EO) => {
    this.setState({mode:4})
  }

  topicDelete = (id) => {
    this.setState( {deleteCode:id}, this.delete);
  }

  editTopic = (id) => {
    this.setState( {editCode:id, mode:3})
  }
  
  render() {
    if ( !this.state.dataReady )
      return <div>загрузка данных...</div>;

    let topics = [...this.state.topics];
    let topics2 = [...this.state.topics];

    var topicsCode=topics.map( topic =>
      <Topic key={topic.id} 
      mode={this.state.mode}
      topic={topic}  />
    );

    var editCode = topics2.find( topic => topic.id === this.state.editCode); 

    let idNum = Math.random();

    return (
      <div>
        <div>Topics</div>
        <table>
        <thead>
              <tr>
                  <th>Title</th>
                  <th>Key words</th>
                  <th>Author</th>
                  <th>Info</th>
                  <th>Edit</th>
                  <th>Delete</th>
              </tr>
          </thead>
          <tbody>
          {topicsCode}
          </tbody>
        </table>
        <input type="button" value="Add a new topic" onClick = {this.addTopic} disabled = {(this.state.mode===4)?true:false}/>
        {
        (this.state.mode===4)&&
          <AddTopic key={idNum}
          mode={this.state.mode}
          id={idNum}
          title=''
          mainWords=''
          author=''
          info=''
        />
        }
        {
        (this.state.mode===3)&&
          <EditSpeaker key={editCode.id}
          mode={this.state.mode}
          id={editCode.id}
          title= {editCode.title}
          mainWords={editCode.mainWords}
          author={editCode.author}
          info={editCode.info}/>
        }
      
      </div>
    )
    ;

  }

}

export default Topics;
