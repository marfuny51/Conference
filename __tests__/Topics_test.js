import React from 'react'
import Enzyme, { shallow, render, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json'
import Topic from '../components/Topic'
import Topics from '../components/Topics'

Enzyme.configure({ adapter: new Adapter() });

const component = shallow(<Topics />);
const topics = [{id:3, title:'How to be happy?', mainWords: 'Happiness', author: 'Ivan Ivanov', info:'Do not worry, be happy'}, {id:4, title:'How to be a programmer?', mainWords: 'IT-industry', author: 'Pavel Pashkin', info:'Anyone can become a programmer'}]
describe('<Topics /> component', () => {
    it('should render', () => {
        component.setState({ dataReady: true, topics: topics});
        topics.map( topic =>
            shallow(<Topic key={topic.id} 
            topic={topic}/>)
          );
        const tree = toJson(component);
        expect(tree).toMatchSnapshot();

        component.find('.Add').simulate("click");

    });
});