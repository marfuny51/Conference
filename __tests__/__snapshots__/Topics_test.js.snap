// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Topics /> component should render 1`] = `
<div
  className="Table"
>
  <table
    className="TableTopic"
  >
    <colgroup>
      <col
        className="col1"
      />
      <col
        className="col2"
        span="2"
      />
      <col
        className="col3"
      />
      <col
        className="col4"
        span="2"
      />
    </colgroup>
    <thead
      className="TableHead"
    >
      <tr>
        <th>
          Title
        </th>
        <th>
          Key words
        </th>
        <th>
          Author
        </th>
        <th>
          Info
        </th>
        <th>
          Edit
        </th>
        <th>
          Delete
        </th>
      </tr>
    </thead>
    <tbody
      className="TableBody"
    >
      <Topic
        key="3"
        mode={null}
        readId={null}
        topic={
          Object {
            "author": "Ivan Ivanov",
            "id": 3,
            "info": "Do not worry, be happy",
            "mainWords": "Happiness",
            "title": "How to be happy?",
          }
        }
      />
      <Topic
        key="4"
        mode={null}
        readId={null}
        topic={
          Object {
            "author": "Pavel Pashkin",
            "id": 4,
            "info": "Anyone can become a programmer",
            "mainWords": "IT-industry",
            "title": "How to be a programmer?",
          }
        }
      />
    </tbody>
  </table>
  <input
    className="Add"
    disabled={false}
    onClick={[Function]}
    type="button"
    value="Add a new topic"
  />
</div>
`;
